src/runtime/storageNG/reference-mode-store.ts

```diff
@@ -0,0 +1,598 @@
+                         ReferenceContainer extends CRDTSingletonTypeRecord<Reference> | CRDTCollectionTypeRecord<Reference>,
+                         Container extends CRDTSingletonTypeRecord<Entity> | CRDTCollectionTypeRecord<Entity>>(
+      backingKey: StorageKey, storageKey: StorageKey, exists: Exists, type: Type, backingConstructor: new () => CRDTModel<CRDTEntityTypeRecord<S, C>>,
+      referenceConstructor: new () => CRDTModel<ReferenceContainer>, modelConstructor: new () => CRDTModel<Container>) {
+    const result = new ReferenceModeStore<Entity, S, C, ReferenceContainer, Container>(storageKey, exists, type, StorageMode.ReferenceMode, modelConstructor);
+    result.backingStore = await BackingStore.construct(backingKey, exists, type, StorageMode.Backing, backingConstructor);
+    result.containerStore = await DirectStore.construct(storageKey, exists, type, StorageMode.Direct, referenceConstructor);
+    result.registerStoreCallbacks();
+    return result;
+  }
+
+  on(callback: ProxyCallback<Container>): number {
+    const id = this.nextCallbackID++;
+    this.callbacks.set(id, callback);
+    return id;
+  }
+  off(callback: number) {
+    this.callbacks.delete(callback);
+  }
+
+  private registerStoreCallbacks() {
+    this.backingStore.on(this.onBackingStore.bind(this));
+    this.containerStore.on(this.onContainerStore.bind(this));
+  }
+
+  /**
+   * Messages are enqueued onto an object-wide queue and processed in order. 
+   * Internally, each handler (handleContainerStore, handleBackingStore, handleProxyMessage)
+   * should not return until the response relevant to the message has been received.
+   * 
+   * When handling proxy messages, this implies 2 rounds of update - first the backing
+   * store needs to be updated, and once that has completed then the container store needs
+   * to be updated.
+   */
+  async onContainerStore(message: ProxyMessage<ReferenceContainer>) {
+    return this.enqueue({from: 'container', message});
+  }
+
+  async onBackingStore(message: ProxyMessage<CRDTEntityTypeRecord<S, C>>, muxId: string) {
+    return this.enqueue({from: 'backing', message, muxId});
+  }
+
+  async onProxyMessage(message: ProxyMessage<Container>): Promise<boolean> {
+    return this.enqueue({from: 'proxy', message});
+  }
+
+  /**
+   * enqueue an incoming update onto the object-wide queue and return a promise that will be resolved
+   * when the update is processed.
+   */
+  private async enqueue(entry: PreEnqueuedMessage<Container, CRDTEntityTypeRecord<S, C>, ReferenceContainer>): Promise<boolean> {
+    return new Promise((resolve, reject) => {
+      const startProcessing = this.receiveQueue.length === 0;
+      this.receiveQueue.push({...entry, promise: resolve});
+      if (startProcessing) {
+        void this.processQueue();
+      }
+    });
+  }
+
+  private async processQueue() {
+    while (this.receiveQueue.length > 0) {
+      const nextMessage = this.receiveQueue[0];
```

Comment:
FYI you can do this instead: `const nextMessage = this.receiveQueue.shift();` THis way there's no danger of forgetting to shift the array at the end of the loop.

Reply:
Yeah, that's correct; can't shift until the end because otherwise the signalling gets messed up. I'll definitely add a comment here.